design/secret_rotation_design.md

```diff
@@ -0,0 +1,951 @@
+    - Changes to the data field of a source Kubernetes Secret (`k8s-secret` destination mode)
+
+    **User Story:** US-03
+    
+  - **Use of SHA-256 checksums to Detect Changes:**<br />
+    For added security, the Secrets Provider will not cache any unencrypted,
+    retrieved secret values in order to detect when Kubernetes Secrets or
+    Push-to-File secret files need to be updated. Instead, the Secrets Provider
+    will store a SHA-256 checksum of the last rendered content that was written
+    to each target Kubernetes Secret or secret file for detecting when contents
+    have changed.
+
+    **User Story:** N/A
+
+- **Deprecation of Env-Variable-Based Configuration for `k8s_secrets` with `init`/`sidecar` modes**
+
+  The configuration of Secrets Provider using environment variables is
+  deprecated for `k8s-secret` destination mode when the Secrets Provider
+  is configured for `init` or `sidecar` container modes. The user is expected
+  to configure the Secrets Provider using Pod Annotations in these scenarios.
+  When the Secrets Provider is configured using environment variables
+  in these scenarios, deprecation warnings will be dispayed in the Secrets
+  Provider container logs.
+
+  **User Story:** US-05, US-06
+
+- **Future Release: Restart of Applications Following Secret Changes:**
+
+  In some cases, there may be an application that is using the Secrets Provider
+  to access Conjur secrets, but the application is designed to consume the
+  secret values only at startup. Some examples of this scenario are:
+
+  - An application using Push-to-File to create startup configuration files
+  - An application using Push-to-File to create `.env` files
+  - An application consuming Kubernetes Secret content as environment variables
+
+  In these cases, the Secrets Provider can update the target Kubernetes
+  Secrets or Push-to-File secret files, but the application will not
+  immediately consume the updated Conjur secret values unless/until the
+  application is restarted.
+
+  In order to support these use cases, the Secrets Provider supports
+  an option to send a configurable restart signal (e.g. SIGKILL,
+  SIGTERM, etc.) to the application process(es) after target Kubernetes Secrets
+  or secret files have been updated.
+
+  The diagram below shows how application restart works:
+
+  **User Story:** N/A
+
+  ![Application Restart](./how-app-restart-works.png)
+
+#### How to Configure Secret Rotation
+
+The Secrets Provider Secret Rotation feature is configured on an application
+as follows:
+
+1. Start by configuring the application Deployment as described in the
+   [Secrets Provider documentation](https://docs.cyberark.com/Product-Doc/OnlineHelp/AAM-DAP/11.2/en/Content/Integrations/Kubernetes_deployApplicationsConjur-k8s-Secrets.htm?tocpath=Integrations%7COpenShift%252C%20Kubernetes%252C%20and%20GKE%7CDeploy%20Applications%7C_____3),
+
+1. If you would like to use the Secrets Provider as a sidecar container
```

Comment:
If the SP is configured as a Sidecar container and the container fails/crashes - do we need readiness/liveness probes to restart container/pod?